File: 백준/Silver/3184. 양/양.java
import java.io.*;
import java.util.*;


public class Main {
    static int x,y,nx,ny;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());
        char[][] arr = new char[N][M];
        Boolean[][] visited = new Boolean[N][M];
        for (int i = 0; i < N; i++) {
            String lst = br.readLine();
            for (int j = 0; j < M; j++) {
                arr[i][j] = lst.charAt(j);
            }
        }
        for (int i = 0; i < N; i++) {
            Arrays.fill(visited[i], false);
        }
        int[][] dir = {{1, 0}, {0, 1}, {-1, 0}, {0, -1}};
        int ansO = 0, ansV=0;
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                if (visited[i][j]) {continue;}
                int o = 0, v = 0;
                if (arr[i][j] == 'o') {o+=1;}
                else if (arr[i][j] == 'v') {v+=1;}
                else{continue;}
                Queue<int[]> q = new ArrayDeque<>();
                q.add(new int[]{i, j});
                visited[i][j] = true;
                while (!q.isEmpty()) {
                    int[] info = q.poll();
                    x = info[0];
                    y = info[1];
                    for (int d = 0; d < 4; d++) {
                        nx = dir[d][0] + x;
                        ny = dir[d][1] + y;
                        if (0 > nx || nx>=N || 0>ny || ny>=M) {continue;}
                        if (arr[nx][ny] == '#') {continue;}
                        if (visited[nx][ny]){continue;}
                        visited[nx][ny]=true;
                        q.add(new int[]{nx,ny});
                        if (arr[nx][ny] == 'o') {o+=1;}
                        else if (arr[nx][ny] == 'v') {v+=1;}
                    }
                }
                if (o > v) {ansO+=o;}
                else {ansV += v;}


            }
        }
        bw.write(ansO+" "+ansV);
        bw.close();
    }
}

File: 백준/Silver/3184. 양/양.py
from collections import deque

def Main():
    N,M = map(int,input().split())
    arr = [list(input()) for i in range(N)]
    visited = [[False]*M for i in range(N)]
    ansO=0
    ansV=0
    for i in range(N):
        for j in range(M):
            if visited[i][j]: continue
            v = 0
            o=0
            if arr[i][j] == "v":
                v+=1
            elif arr[i][j] == "o":
                o+=1
            else:continue
            q = deque()
            q.append((i,j))
            visited[i][j]=True
            while q :
                x,y = q.popleft()
                for dx,dy in [(1,0),(-1,0),(0,1),(0,-1)]:
                    nx,ny=dx+x,dy+y
                    if 0<=nx<N and 0<=ny<M and arr[nx][ny] != "#" and not visited[nx][ny]:
                        visited[nx][ny] = True
                        q.append((nx,ny))
                        if arr[nx][ny] == "v":v+=1
                        elif arr[nx][ny] == "o":o+=1
            if o>v:ansO+=o
            else:ansV+=v
    
    print(ansO,ansV)
Main()